oio/account/backend.py

```diff
@@ -202,83 +204,115 @@ def update_container(self, account_id, name, mtime, dtime, object_count,
+               "local name = ARGV[1];\n"
+               "local mtime = redis.call('HGET', KEYS[2], 'mtime');\n"
+               "local dtime = redis.call('HGET', KEYS[2], 'dtime');\n"
+               "local bytes = redis.call('HGET', KEYS[2], 'bytes');\n"
+
+               # When the keys do not exist redis return false and not nil
+               "if objects == false then\n"
+               "  objects = 0\n"
+               "end\n"
+               "if dtime == false then\n"
+               "  dtime = '0'\n"
+               "end\n"
+               "if mtime == false then\n"
+               "  mtime = '0' \n"
+               "end\n"
+               "if bytes == false then\n"
+               "  bytes = 0\n"
+               "end\n"
+
+               "local old_mtime = mtime;\n"
+               "local inc_objects;\n"
+               "local inc_bytes;\n"
+
+               "if not is_sup(ARGV[3],dtime) and "
+               "   not is_sup(ARGV[2],mtime) then\n"
+               "  return redis.error_reply('no_update_needed');\n"
+               "end;\n"
+
+               "if is_sup(ARGV[2],mtime) then\n"
+               "  mtime = ARGV[2];\n"
+               "end;\n"
+
+               "if is_sup(ARGV[3],dtime) then\n"
+               "  dtime = ARGV[3];\n"
+               "end;\n "
+               "if is_sup(dtime,mtime) then\n"
+               "  inc_objects = -objects;\n"
+               "  inc_bytes = -bytes;\n"
+               "  redis.call('HMSET', KEYS[2], 'bytes', 0, 'objects', 0);\n"
+               "  redis.call('EXPIRE', KEYS[2], tonumber(ARGV[8]));\n"
+               "  redis.call('ZREM', KEYS[3], name);\n"
+               "elseif is_sup(mtime,old_mtime) then\n"
+               "  redis.call('PERSIST', KEYS[2]);\n"
+               "  inc_objects = tonumber(ARGV[4]) - objects\n"
+               "  inc_bytes = tonumber(ARGV[5]) - bytes\n"
+               "  redis.call('HMSET', KEYS[2], 'bytes', tonumber(ARGV[5]),\n"
+               "             'objects', tonumber(ARGV[4]));\n"
+               "  redis.call('ZADD', KEYS[3], '0', name);\n"
+               "else"
+               "  return redis.error_reply('no_update_needed');\n"
+               "end;\n"
+
+               "redis.call('HMSET', KEYS[2], 'mtime', mtime,"
+               "           'dtime', dtime, 'name', name)\n"
+               "if inc_objects ~= 0 then\n"
+               "  redis.call('HINCRBY', KEYS[4], 'objects', inc_objects);\n"
+               "end;\n"
+               "if inc_bytes ~= 0 then\n"
+               "  redis.call('HINCRBY', KEYS[4], 'bytes', inc_bytes);\n"
+               "end;\n"
+               )
+
+        keys = [account_id, AccountBackend.ckey(account_id, name),
+                ("containers:%s" % (account_id)),
+                ("account:%s" % (account_id))]
+        args = [name, mtime, dtime, object_count, bytes_used,
+                self.autocreate, Timestamp(time()).normal, EXPIRE_TIME]
+
+        update = conn.register_script(lua)
```

Comment:
What is the cost of registering script for each update_container call  ?
Could it be cached on `self` lifetime ?